README
Cisco-AccessList-Parser version 0.0.1

[ REPLACE THIS...

  The README is used to introduce the module and provide instructions on
  how to install the module, any machine dependencies it may have (for
  example C compilers and installed libraries) and any other information
  that should be understood before the module is installed.

  A README file is required for CPAN modules since CPAN extracts the
  README file from a module distribution so that people browsing the
  archive can use it get an idea of the modules uses. It is usually a
  good idea to provide version information here so that people can
  decide whether fixes for the module are worth downloading.
]


INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

Alternatively, to install with Module::Build, you can use the following commands:

	perl Build.PL
	./Build
	./Build test
	./Build install


DEPENDENCIES

None.


COPYRIGHT AND LICENCE

Copyright (C) 2011, stereocat

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.


Changes
Revision history for Cisco-AccessList-Parser

0.0.1  Tue Sep 13 22:35:17 2011
       Initial release.



t/test-runner.pl
#!/usr/bin/perl

use strict;
use warnings;
use YAML::XS;
use Data::Dumper;
use Test::More;
use Parse::Eyapp;
use Cisco::AccessList::Parser;

if ( !defined( $ARGV[0] ) ) {
    die "usage: $0 test-data.dat";
}
my $test_data_output_file = $ARGV[0];
my $test_case             = YAML::XS::LoadFile($test_data_output_file)
    or die "Cannot load $test_data_output_file";

my $test_count = @$test_case;
plan (tests => $test_count);
my $p = Cisco::AccessList::Parser->new();

foreach my $test (@$test_case) {
    my $data  = $test->{data};
    my $valid = $test->{valid};
    my $todo  = $test->{todo};

    $p->set_yydata_input($data);

    if ( !( defined($todo) && length($todo) > 0 ) ) {
        cmp_ok($p->is_acl_accepted(), '==', $valid, $data);
    }
    else {
      TODO: {
            local $TODO = $todo;
            $p->set_yydata_input($data);
            cmp_ok($p->is_acl_accepted(), '==', $valid, $data);
        }
    }
}


generate_tests.PL
#!/usr/bin/perl
use strict;
use warnings;
use YAML;
use Path::Class;

sub generate_case {
    my $field_data = shift;
    my $fieldseq   = shift;

    my $test_field = shift @$fieldseq;

    if ( !defined($test_field) ) {
        return [ { data => "", valid => "1", todo => undef } ];
    }
    if ( !exists( $field_data->{$test_field} ) ) {
        die "Unknown field name: $test_field";
    }

    my $later_test_case = generate_case( $field_data, $fieldseq );

    my $result = [];
    foreach my $field ( @{ $field_data->{$test_field} } ) {
        my $curr_data  = $field->{data};
        my $curr_valid = $field->{valid};
        my $curr_todo  = $field->{todo};

        foreach my $later_field (@$later_test_case) {

            my $later_data  = $later_field->{data};
            my $later_valid = $later_field->{valid};
            my $later_todo  = $later_field->{todo};

            my $data = join( q{ }, $curr_data, $later_data );
            $data =~ s/\s+$//;
            my $valid = $curr_valid * $later_valid;
            my $todo = defined($curr_todo) ? $curr_todo : $later_todo;

            push( @$result,
                { data => $data, valid => $valid, todo => $todo } );
        }
    }
    return $result;
}
############################################################

sub generate_test_script {
    my $code                 = shift;
    my $field_data_dir       = shift;
    my $test_name            = shift;
    my $generate_test_script = shift;
    my $test_script_file     = file( $field_data_dir, $test_name . '.t' );
    open my $tsfh, ">", $test_script_file
        or die "Cannot open $test_script_file: $!";

    print $tsfh $code;

    foreach my $testcase (@$generate_test_script) {
        my $data  = $testcase->{'data'};
        my $todo  = defined( $testcase->{'todo'} ) ? $testcase->{'todo'} : "";
        my $valid = $testcase->{'valid'};
        my $record = <<"EOL";
===
--- data: $data
--- todo: $todo
--- valid: $valid
EOL
        print $tsfh $record;
    }
    close $tsfh;
    print "# Generate Test Script: $test_script_file\n";
}

############################################################

# read __DATA__ token; test running program
my $test_runner_code = do { local $/; <DATA> };

foreach my $test_conf_file (<t/*.conf>) {
    my $test_conf_list = YAML::LoadFile($test_conf_file)
        or die "Cannot Load Test Conf File: $test_conf_file";
    print "# Load Test Conf File: $test_conf_file\n";

    foreach my $test_case (@$test_conf_list) {

        # data check
        if ( !exists( $test_case->{testname} ) ) {
            die "test case name: not specified in conf file: $test_conf_list";
        }
        if ( !exists( $test_case->{casedata} ) ) {
            die "data file: not specified in conf file: $test_conf_file";
        }
        if ( !exists( $test_case->{fieldseq} ) ) {
            die "fieldseq-data: not contained in conf file: $test_conf_file";
        }

        # data construction sequence data list
        my $fieldseq = $test_case->{fieldseq};

        # load data file (input)
        my $field_data_file = $test_case->{casedata};
        my $field_data_dir  = file($test_conf_file)->parent;
        my $field_data
            = YAML::LoadFile( file( $field_data_dir, $field_data_file ) );

        # generated test case
        my $generated_test_case = generate_case( $field_data, $fieldseq );

        # save generated test case to dat file
        generate_test_script(
            $test_runner_code,      $field_data_dir,
            $test_case->{testname}, $generated_test_case
        );
    }
}

__DATA__
#!/usr/bin/perl

use strict;
use warnings;
use Test::Base;
use Test::Warn;
use Cisco::AccessList::Parser;

my $p = Cisco::AccessList::Parser->new();

plan tests => 1 * blocks;

sub check_acl {
    my $p = shift;
    my $block = shift;

    my $data  = $block->data();
    my $valid = $block->valid();
    my $todo  = $block->todo();
    my ($acl, $objgrp);

    if($valid > 0) {
        ( $acl, $objgrp ) = $p->parse( 'input' => $data );
        is( keys(%$acl) + keys(%$objgrp), $valid, $data );
    }
    else {
        warnings_exist {
            ( $acl, $objgrp ) = $p->parse( 'input' => $data );
        } qr/WARNING:/, $data;
    }
}

run {
    my $block = shift;
    my $data  = $block->data();
    my $valid = $block->valid();
    my $todo  = $block->todo();

    if ( !( defined($todo) && length($todo) > 0 ) ) {
        check_acl($p, $block);
    }
    else {
      TODO: {
        local $TODO = $todo;
        check_acl($p, $block);
      }
    }
};
__DATA__


t/testcase-generator.pl
#!/usr/bin/perl
use strict;
use warnings;
use YAML;
use Path::Class;

sub generate_case {
    my $field_data = shift;
    my $fieldseq   = shift;

    my $test_field = shift @$fieldseq;

    if ( !defined($test_field) ) {
        return [ { data => "", valid => "1", todo => undef } ];
    }
    if ( !exists( $field_data->{$test_field} ) ) {
        die "Unknown field name: $test_field";
    }

    my $later_test_case = generate_case( $field_data, $fieldseq );

    my $result = [];
    foreach my $field ( @{ $field_data->{$test_field} } ) {
        my $curr_data  = $field->{data};
        my $curr_valid = $field->{valid};
        my $curr_todo  = $field->{todo};

        foreach my $later_field (@$later_test_case) {

            my $later_data  = $later_field->{data};
            my $later_valid = $later_field->{valid};
            my $later_todo  = $later_field->{todo};

            my $data = join( q{ }, $curr_data, $later_data );
            $data =~ s/\s+$//;
            my $valid = $curr_valid * $later_valid;
            my $todo = defined($curr_todo) ? $curr_todo : $later_todo;

            push( @$result,
                { data => $data, valid => $valid, todo => $todo } );
        }
    }
    return $result;
}
############################################################

sub generate_test_script {
    my $code                 = shift;
    my $field_data_dir       = shift;
    my $test_name            = shift;
    my $generate_test_script = shift;
    my $test_script_file     = file( $field_data_dir, $test_name . '.t' );
    open my $tsfh, ">$test_script_file"
        or die "Cannot open $test_script_file";

    print $tsfh $code;

    foreach my $testcase (@$generate_test_script) {
        my $data   = $testcase->{'data'};
        my $todo   = defined( $testcase->{'todo'} ) ? $testcase->{'todo'} : 0;
        my $valid  = $testcase->{'valid'};
        my $record = <<"EOL";
===
--- data: $data
--- todo: $todo
--- valid: $valid
EOL
        print $tsfh $record;
    }
    close $tsfh;
}
############################################################

# read test-configuration file
die "usage: $0 test.conf [test.conf] ..." if ( @ARGV < 1 );

# read __DATA__ token; test running program
my $test_runner_code = do { local $/; <DATA> };

foreach my $test_conf_file (@ARGV) {
    my $test_conf_list = YAML::LoadFile($test_conf_file);

    foreach my $test_case (@$test_conf_list) {

        # data check
        if ( !exists( $test_case->{testname} ) ) {
            die "test case name: not specified in conf file: $test_conf_list";
        }
        if ( !exists( $test_case->{casedata} ) ) {
            die "data file: not specified in conf file: $test_conf_file";
        }
        if ( !exists( $test_case->{fieldseq} ) ) {
            die "fieldseq-data: not contained in conf file: $test_conf_file";
        }

        # data construction sequence data list
        my $fieldseq = $test_case->{fieldseq};

        # load data file (input)
        my $field_data_file = $test_case->{casedata};
        my $field_data_dir  = file($test_conf_file)->parent;
        my $field_data
            = YAML::LoadFile( file( $field_data_dir, $field_data_file ) );

        # generated test case
        my $generated_test_case = generate_case( $field_data, $fieldseq );

        # save generated test case to dat file
        generate_test_script(
            $test_runner_code,      $field_data_dir,
            $test_case->{testname}, $generated_test_case
        );
    }
}

__DATA__
#!/usr/bin/perl

use strict;
use warnings;
use Test::Base;
use Cisco::AccessList::Parser;

my $p = Cisco::AccessList::Parser->new();

plan tests => 1 * blocks;

run {
    my $block = shift;
    my $data  = $block->data();
    my $valid = $block->valid();
    my $todo  = $block->todo();

    $p->set_yydata_input($data);

    if ( !( defined($todo) && length($todo) > 0 ) ) {
        is( $p->is_acl_accepted(), $valid, $data );
    }
    else {
    TODO: {
            local $TODO = $todo;
            $p->set_yydata_input($data);
            is( $p->is_acl_accepted(), $valid, $data );
        }
    }
};
__DATA__


check2.pl
# -*- cperl -*-

use lib qw(.);
use strict;
use warnings;

use Parse::Eyapp;

use Data::Dumper;
use Cisco::AccessList::Parser;

#sub TERMINAL::info { $_[0]{attr} }

my $input = << "EOACL";
ip access-list extended FA0-IN
 deny   ip 10.0.0.0 0.255.255.255 any log
 deny   ip 172.16.0.0 0.15.255.255 any log
 deny   ip 192.168.0.0 0.0.255.255 any log
 deny   ip 0.0.0.0 0.255.255.255 any log
 deny   ip 127.0.0.0 0.255.255.255 any log
 deny   ip 192.0.2.0 0.0.0.255 any log
 deny   ip 169.254.0.0 0.0.255.255 any log
 deny   ip 224.0.0.0 31.255.255.255 any log
 deny   tcp any any eq 135 log
 deny   udp any any eq 135 log
 deny   tcp any any range 137 139 log
 deny   udp any any range netbios-ns netbios-ss log
 deny   tcp any any eq 445 log
 deny   udp any any eq 445 log
 deny   tcp any any eq 6000 log
 deny   tcp any any eq 1433 log
 remark vpn
 permit udp any eq domain 122.219.206.8 0.0.0.7
 remark vpn
 permit esp any any
 permit tcp any any eq 50
 permit tcp any any eq 51
 permit udp any any eq isakmp
 permit udp any any eq 1701
 remark ntp
 permit udp host 210.188.224.14 eq ntp any
 permit udp host 61.122.112.135 eq ntp any
 remark 6to4
 permit ip 192.88.99.0 0.0.0.255 any
 permit 41 any any
 remark permit USEN DHCP
 permit udp any eq bootps any eq bootpc
 remark home web server
 permit tcp any any eq 8000
 remark share
 permit tcp any any eq 25010
 remark permit any from inside to outside
 evaluate iptraffic
 permit tcp any any established
 permit icmp any any
 deny   ip any any log
EOACL

my $multiple_std_acl = << 'EOACL';
access-list 1 permit 192.168.0.0 0.0.255.255
access-list 1 deny   any log
access-list 9 permit 192.168.0.0 0.0.255.255
access-list 10 deny   any log
access-list 99 permit 192.168.0.0 0.0.255.255
access-list 99 deny   any log
EOACL

my $multiple_ext_acl = << 'EOACL';
access-list 100 remark DNS Exempt
access-list 100 deny   udp 122.219.206.8 0.0.0.7 any eq domain
access-list 100 deny   tcp 122.219.206.8 0.0.0.7 any eq domain
access-list 100 remark NTP
access-list 100 permit tcp any host 210.197.74.200
access-list 100 permit udp any eq ntp any eq ntp
access-list 100 remark 6to4
access-list 100 permit 41 any host 192.88.99.1
access-list 100 permit ip any host 192.88.99.1
access-list 100 remark others
access-list 100 permit tcp any eq 0 any eq 0
access-list 100 permit udp any eq 0 any eq 0
access-list 100 deny   ip any any log
access-list 110 remark SPLIT_VPN
access-list 110 permit ip 192.168.0.0 0.0.255.255 any
access-list 120 permit ip 192.168.0.0 0.0.255.255 any
access-list 120 permit tcp any any log
EOACL

my $multiple_std_ext_acl_mix = << 'EOACL';
access-list 1 permit 192.168.0.0 0.0.255.255
access-list 1 deny   any log
access-list 9 permit 192.168.0.0 0.0.255.255
access-list 10 deny   any log
access-list 99 permit 192.168.0.0 0.0.255.255
access-list 99 deny   any log
access-list 100 remark DNS Exempt
access-list 100 deny   udp 122.219.206.8 0.0.0.7 any eq domain
access-list 100 deny   tcp 122.219.206.8 0.0.0.7 any eq domain
access-list 100 remark NTP
access-list 100 permit tcp any host 210.197.74.200
access-list 100 permit udp any eq ntp any eq ntp
access-list 100 remark 6to4
access-list 100 permit 41 any host 192.88.99.1
access-list 100 permit ip any host 192.88.99.1
access-list 100 remark others
access-list 100 permit tcp any eq 0 any eq 0
access-list 100 permit udp any eq 0 any eq 0
access-list 100 deny   ip any any log
access-list 110 remark SPLIT_VPN
access-list 110 permit ip 192.168.0.0 0.0.255.255 any
access-lst 110 permit hoge
access-list 110 deny ip any any log
access-list 120 permit ip 192.168.0.0 0.0.255.255 any
access-list 120 permit tcp any any log
EOACL

my $multiple_named_std_acl = << 'EOACL';
ip access-list standard remote-ipv4
 permit 192.168.0.0 0.0.255.255
 remark deny all
 deny   any log
!
ip access-list standard test-acl
 permit 192.168.0.0 0.0.255.255
 permit 192.168.2.0 0.0.255.255
 permit host 192.168.4.5
 deny   any log
!
EOACL

my $multiple_named_ext_acl = << 'EOACL';
!
ip access-list extended FA0-IN
 deny   ip 169.254.0.0 0.0.255.255 any log
 deny   ip 224.0.0.0 31.255.255.255 any log
 deny   tcp any any eq 135 log
!
ip access-list extended FA0-OUT
 deny   ip any 10.0.0.0 0.255.255.255 log
 permit icmp any any
 permit ip any any reflect iptraffic timeout 300
 deny   ip any any log
!
EOACL

my $multiple_named_ext_std_acl_mix = << 'EOACL';
ip access-list extended FA0-IN
 deny   ip 169.254.0.0 0.0.255.255 any log
 deny   ip 224.0.0.0 31.255.255.255 any log
 deny   tcp any any eq 135 log
!
ip access-list extended FA0-OUT
 deny   ip any 10.0.0.0 0.255.255.255 log
 permit icmp any any
 permit ip any any reflect iptraffic timeout 300
 deny   ip any any log
!
ip access-list extended FA0-IN
 deny   ip 169.254.0.0 0.0.255.255 any log
 deny   ip 224.0.0.0 31.255.255.255 any log
 deny   tcp any any eq 135 log
!
ip access-list extended FA0-OUT
 deny   ip any 10.0.0.0 0.255.255.255 log
 permit icmp any any
 permit ip any any reflect iptraffic timeout 300
 deny   ip any any log
!
EOACL

my $input2 = << 'EOACL';
ip access-list standard remote-ipv4
 permit 192.168.0.0 0.0.255.255
 deny   any log
!
ip access-list extended FA0-IN
 deny   ip 169.254.0.0 0.0.255.255 any log
 deny   ip 224.0.0.0 31.255.255.255 any log
 deny   tcp any any eq 135 log
!
ip access-list extended FA0-OUT
 deny   ip any 10.0.0.0 0.255.255.255 log
 permit icmp any any
 permit ip any any reflect iptraffic timeout 300
 deny   ip any any log
!
EOACL

my $input3 = << 'EOACL';
access-list 1 permit 192.168.0.0 0.0.255.255
access-list 1 deny   any log
access-list 100 remark DNS Exempt
access-list 100 deny   udp 122.219.206.8 0.0.0.7 any eq domain
access-list 100 deny   tcp 122.219.206.8 0.0.0.7 any eq domain
access-list 100 remark NTP
access-list 100 permit tcp any host 210.197.74.200
access-list 100 permit udp any eq ntp any eq ntp
access-list 100 remark 6to4
access-list 100 permit 41 any host 192.88.99.1
access-list 100 permit ip any host 192.88.99.1
access-list 100 remark others
access-list 100 permit tcp any eq 0 any eq 0
access-list 100 permit udp any eq 0 any eq 0
access-list 100 deny   ip any any log
access-list 110 remark SPLIT_VPN
access-list 110 permit ip 192.168.0.0 0.0.255.255 any
EOACL

my $input5 = << 'EOACL';
access-list 130 deny   ip any any log
access-list 130 remark SPLIT_VPN
access-list 130 permit ip 192.168.0.0 0.0.255.255 any
access-list 120 remark SPLIT_VPN
access-list 120 permit udp any eq 0 any eq 0
access-list 100 permit udp any eq 0 any eq 0
access-list 100 deny   ip any any log
access-list 100 permit ip 192.168.0.0 0.0.255.255 any
EOACL

my $input6 = << 'EOACL';
ip access-list standard remote-ipv4
 10 deny host 192.168.0.33 log
 20 permit 192.168.0.0 0.0.255.255
 30 deny   any log
EOACL

my $input7 = << "EOACL";
ip access-list extended FA0-IN
 deny   tcp any any eq 1433 log
 remark ip vpn
 permit udp any eq domain 122.219.206.8 0.0.0.7
 deny   ipinip any any log
 deny   ip any any log
EOACL

my $input8 = << "EOACL";
ip access-list standard remote-ipv4
 deny host 192.168.0.33 log
 permit 192.168.0.0 0.0.255.255
 deny   any any log
EOACL

my $input9 = << "EOACL";
ip access-list standard remote-ipv4
 deny host 192.168.0.33 log
 permit 192.168.0.0 0.0.255.255
 deny   any log
EOACL

my $objgrp1 = << "EOACL";
object-group network SMTP_Server
 description ISP SMTP server
 host 192.168.0.2
 192.168.1.0 /24
 192.168.2.0 0.0.0.128
 group-object other_network
 host 172.16.2.3
 range 192.168.3.2 192.168.3.200
!
object-group service Web_Service
 description web service
 icmp echo
 tcp smtp
 tcp telnet
 tcp source range 1 65535
 udp domain
 tcp-udp range 2000 2005
 group-object other-list
EOACL

my $test_data = $input2;
my $aclparser = Cisco::AccessList::Parser->new();
my $debug     = 0x1F;
my ( $acl, $objgrp )
    = $aclparser->parse( 'input' => $test_data, 'debug' => $debug );
print "RESULT: ", ( keys(%$acl) + keys(%$objgrp) ), "\n";
print "==================\n";
$aclparser->lex_check( 'input' => $test_data );



check3.pl
# -*- cperl -*-

use lib qw(.);
use strict;
use warnings;

use Parse::Eyapp;
use Test::More;
use Test::Warn;
use Data::Dumper;
use Cisco::AccessList::Parser;

#sub TERMINAL::info { $_[0]{attr} }

plan tests => 2;

my $input = << "EOACL";
access-list 100 permit 256 192.168.0.0 0.0.0.128 any
EOACL

my $aclparser = Cisco::AccessList::Parser->new();

my $data = $input;
my $debug     = 0;

my ( $acl, $objgrp );
warning_like {
    ( $acl, $objgrp ) = $aclparser->parse( 'input' => $data );
} [qr/WARNING: Protocol/, qr/WARNING: Syntax error/], $data;

warnings_exist {
    ( $acl, $objgrp ) = $aclparser->parse( 'input' => $data );
} qr/WARNING: Protocol/, $data;

# print "RESULT: ", ( keys(%$acl) + keys(%$objgrp) ), "\n";
# print "==================\n";
# $aclparser->lex_check( 'input' => $test_data );

__END__

access-list 100 permit 256 host 192.168.0.1 10.1.0.2
                       ~~~                           ~~~~~~~~~
複数箇所のエラー: 複数エラーの出力/テストをどうハンドリングするか

----------------------------------------
In state 40:
Stack: 0->'EXTACL'->12->'dynamic_spec'->27->'action'->69->'NUMBER'->40
Don't need token.
Reduce using rule 31 (ip_proto --> NUMBER): WARNING: Syntax error, near: >end-of-input<.
Expected one of these tokens: HOST IPV4_ADDR ANY
Back to state 69, then go to state 229.
----------------------------------------
In state 229:
Stack: 0->'EXTACL'->12->'dynamic_spec'->27->'action'->69->'ip_proto'->229
Need token. Got >HOST<
Shift and go to state 77.
----------------------------------------

数値チェックで yyerror がよばれているけど ip_proto は受理されてしまっている

    


Build.PL
use strict;
use warnings;
use Module::Build;

my $builder = Module::Build->new(
    module_name    => 'Cisco::AccessList::Parser',
    dynamic_config => 1,
    PL_files       => {
        'lib/Cisco/AccessList/Parser.pm.PL' =>
            'lib/Cisco/AccessList/Parser.pm',
        'generate_tests.PL' => [
            't/10-stdacl.t',
            't/20-extacl-icmp.t',
            't/21-extacl-ip.t',
            't/22-extacl-tcp.t',
            't/23-extacl-udp.t',
            't/24-extacl-objgrp1.t',
            't/25-extacl-objgrp2.t',
            't/26-extacl-objgrp3.t',
            't/27-extacl-options1.t',
            't/28-extacl-options2.t'
        ],
    },
    license            => 'perl',
    dist_author        => 'stereocat <stereocat@gmail.com>',
    dist_version_from  => 'lib/Cisco/AccessList/Parser.eyp',
    dist_abstract      => 'Cisco access-list parse',
    configure_requires => { 'Module::Build' => 0.36, },
    requires           => {
        'Hash::Merge' => 0.12,
        'Regexp::Common' => 2011000000,
        'Parse::Eyapp' => 1.181,
        'version'      => 0,
    },
    build_requires => {
        'YAML'       => 0,
        # 'Test::Most' => 0,
        'Test::More' => 0,
        'Test::Base' => 0.60,
    },
    add_to_cleanup => ['Cisco-AccessList-Parser-*'],
);

$builder->create_build_script();

